enhancements/sig-architecture/224-event-based-manifestwork/README.md

```diff
@@ -0,0 +1,788 @@
+
+To facilitate easy adoption of the event-based mechanism, a client interface should be provided, enabling existing work agents or consumers to transition without extensive code changes.
+
+### Goals
+
+- Establish a standardized event protocol for delivering manifestworks and returning status information.
+- Define a resync mechanism to handle failures and ensure synchronization.
+- Develop a client interface that facilitates easy integration for consumers.
+
+### Non-Goals
+
+- Define the controlplane to act as a consumer of the protocol.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+- As a hub controller developer, I can seamlessly transition to event-based manifestworks without modifying the controller logic by switching from the Kubernetes-style approach.
+
+#### Story 2
+
+- As a service developer, I can effortlessly utilize the event-based manifestwork API to construct a service that delivers manifestworks to managed clusters.
+
+### Risks and Mitigation
+
+- The authentication to the event transport layer (TBD)
+
+## Design Details
+
+We will adopt the CloudEvents specification as the message format for ManifestWorks. Multiple API consumers, referred to as "sources," can utilize this format. Each source must have a unique source ID to ensure proper identification and handling of events.
+
+To ensure the uniqueness and immutability of source components, every source, whether it is a consumer on the open cluster management hub cluster or a RESTful service handling resource requests, should possess a unique identifier. For example, the open cluster management hub consumer can generate a source ID by hashing the hub cluster URL and appending a consumer name. Similarly, a RESTful service can select a unique name or generate a unique ID in the associated database for its source identification.
+
+### Spec Event
+
+Sources send their resources along with the message to the work agent, enabling the creation, updating, or deletion of the respective resources on the managed cluster. "Source" is used to refer to the API consumers or components that utilize the CloudEvents specification for ManifestWorks.
+
+Event transfer flow:
+
+1. The work agent subscribes to the spec event topic `/sources/+/clusters/<cluster-id>/manifests` from the broker, for example: `/sources/+/clusters/cluster1/manifests`.
+2. The source receives high level API request and publishes its resources using the spec event to the topic `/sources/<source-id>/clusters/<cluster-id>/manifests`, for example: `/sources/sdbdeavo-mwrs-hub-controller/clusters/cluster1/manifests`.
```

Comment:
These examples are supportable in MQTT Broker and not kafka. And IMHO I think we align with MQTT for optimization. It will be much harder to optimize for both at the same time.

Reply:
We did not use specific features of the MQTT protocol to optimize, we use the cloud events sdk to do this, the sdk hides the protocol details for us,  in fact, we just need a transporter/broker, we can pub/sub message from this transporter,  regardless of the transporter is MQTT/Kafka/AMQ, etc, we may need to describe this part more generic.

Reply:
@skeeey can we look at the code a bit.  In the example that I see here: https://github.com/cloudevents/sdk-go/blob/main/samples/mqtt/receiver/main.go and from Kafka etc - the sdk wrapper is really minimal. We have access to all broker functions - as it should be. Each broker has different config settings and if we do not use them, it's a wastage. 
Having said that, if we want to make it generic, we should think through how would handle it in Kafka at least. For starter Kafka cannot handle `/` in its topics. So that has to change. The session retention in Kafka is different- this may or may not affect the design.  So @skeeey if we need to make it generic, we probably need to do more experimentation with prototypes.

Reply:
@bjoydeep yeah, we will provide a client (it is described in the clients section) that help developers to finish the pub/sub/resync resources, https://github.com/skeeey/work/blob/mqtt/pkg/clients/mqclient/client.go#L33

With this client, developers provide their  own cloud event based clients (https://github.com/skeeey/work/blob/mqtt/pkg/clients/mqclient/client.go#L35), the clients may be from the MQTT, Kafka, etc. , like you mentioned

And considering different protocols may have their own different feature, we also provide the interface https://github.com/skeeey/work/blob/mqtt/pkg/clients/workclient/protocal/context.go to allow developers to build their cloud event context, e.g.
for MQTT, it may be
```
mqttSubClient, mqttPubClient, err := mqttOptions.GetCloudEventClients(ctx, b.clusterName)
mqttCTXGetter := mqttOptions.NewMQTTContextGetter()

ceClient := NewCloudEventClient[*workv1.ManifestWork](b.source,
			mqttSubClient,
			mqttPubClient,
// ....

// pub resources

ceClient.Publish(mqttCTXGetter.Context(...), work) error
``` 

for Kafka, it may be
```
kafkaSubClient, kafkaClient, err := kafkaOptions.GetCloudEventClients(...)
kafkaCTXGetter := kafkaOptions.NewMQTTContextGetter()

ceClient := NewCloudEventClient[*workv1.ManifestWork](b.source,
			kafkaSubClient,
			kafkaPubClient,
// ....

// pub resources

ceClient.Publish(kafkaCTXGetter.Context(...), work) error
```